Master 2154703 l10n be edi bfr

bfr-o opened the pull request:
[IMP] uniformized way to import/export EDI documents

In details, this is achieved by
 - Adding a generic framework module to import/export EDI documents
 - Adapting existing EDI modules to it
 - Fix PyPDF2 to manage multiple embedded files
 - Adding E-FFF standard in a separated module (belgian EDI export)
 - Fix UBL move line price import.

Task ID 2154703

smetl commented on addons/account_facturx/wizard/account_invoice_send.py:
```diff
@@ -0,0 +1,12 @@
+    _inherit = 'account.invoice.send'
+    _description = 'Account Invoice Send'
+
+    xml_type = fields.Selection(selection_add=[('facturx', 'Factur-X')])
```
I don't think this is a good idea. You probably want to manage 2 flows:
- the xml is embedded inside the pdf: do nothing
- the xml is not embedded inside the pdf: send the pdf + the xml to your customer.

smetl commented on addons/account_invoice_export_ubl/models/account_move.py:
```diff
@@ -0,0 +1,73 @@
+        if not self.invoice_date:
+            raise ValueError('Please set an invoice date')
+        if not self.commercial_partner_id:
+            raise ValueError('Please set a commercial_partner_id')
```
The commercial_partner_id is always set on out_invoice/out_refund

smetl commented on addons/account_invoice_export_ubl/models/account_move.py:
```diff
@@ -0,0 +1,73 @@
+        self.ensure_one()
+
+        if not self.invoice_date:
+            raise ValueError('Please set an invoice date')
```
the invoice_date is always set after posting the invoice

smetl commented on addons/account_invoice_export_ubl/models/account_move.py:
```diff
@@ -0,0 +1,73 @@
+        if xml_type == 'ubl21':
+            return {'name': '%s_ubl21.xml' % (self.name.replace('/', '_')), 'content': self._export_invoice_as_ubl()}
+        else:
+            return super(AccountMove, self)._to_xml(xml_type)
```
I think we could do something cleaner like:

```
def _get_edi_format(self):
  return [
    # type, import_check, to_xml_encoder, from_xml_decoder
    ('ubl', '_is_ubl_xml', '_ubl_convert_to_xml', '_ubl_decode_xml'),
    ...
  ]

def _ubl_convert_to_xml():
  return {
    'xml': self._export_invoice_with_ubl(),
    'filename': <blablabla>,
    'embed_to_pdf': True,
  }
```

smetl commented on addons/l10n_be_edi/__manifest__.py:
```diff
@@ -11,7 +11,7 @@
 Belgian e-invoicing uses the UBL 2.0 using the e-fff protocol.
     """,
-    'depends': ['account', 'l10n_be', 'account_facturx'],
+    'depends': ['account', 'l10n_be', 'account_invoice_import'],
```
you can remove the dependency of `account` since this is an existing dependency of `account_invoice_import`

smetl commented on addons/account_facturx/models/account_move.py:
```diff
@@ -224,97 +228,15 @@ def _import_facturx_invoice(self, tree):
-            self.write({'extract_state': 'done'})
+        # OVERRIDE
+        ubl_decoders = [('Factur-X', self._detect_facturx, self._import_facturx_invoice, self._export_as_facturx_xml, True)]
+        return super(AccountMove, self)._get_xml_decoders() + ubl_decoders
```
You know, the size of this kind of sentence could be reduced: `return super()._get_xml_decoders() + ubl_decoders`

smetl commented on addons/account_invoice_import/models/account_move.py:
```diff
@@ -0,0 +1,124 @@
+
+    def _remove_ocr_option(self):
+        if 'extract_state' in self:
+            self.write({'extract_state': 'done'})
```
should be fine to remove this hack if possible

smetl commented on addons/account_invoice_import/wizard/account_invoice_send.py:
```diff
@@ -0,0 +1,36 @@
+                to_attach = [xml_export() for xml_type, xml_checker, xml_import, xml_export, embed_to_pdf in invoice._get_xml_decoders() if not embed_to_pdf]
+
+                for filename, content in to_attach:
+                    attachment_id = self.env['ir.attachment'].search([('name', '=', filename)], limit=1)
```
Since your variable contains a record, its name should be `attachment`. Keep `_id` for integers :)

smetl replied:
Also, your `[('name', '=', filename)]` is weird.. Shouldn't be something like:
```
[('name', '=', filename), ('res_model', '=', 'account.move'), ('res_id', '=', invoice.id)]
```

smetl replied:
Also, don't make a search inside a loop =x

smetl commented on addons/account_invoice_import/wizard/account_invoice_send.py:
```diff
@@ -0,0 +1,36 @@
+                    self.write({'attachment_ids': [(4, attachment_id.id)]})
+
+        res = super(AccountInvoiceSend, self).send_and_print_action()
+        return res
```
`return super().send_and_print_action()`

smetl commented on addons/account_invoice_import/wizard/account_invoice_send.py:
```diff
@@ -0,0 +1,36 @@
+                            'res_id': invoice.id,
+                            'mimetype': 'application/xml'
+                        })
+                    self.write({'attachment_ids': [(4, attachment_id.id)]})
```
`self.attachment_ids |= attachment`

smetl commented on addons/account_invoice_import/wizard/account_invoice_send.py:
```diff
@@ -0,0 +1,36 @@
+                for filename, content in to_attach:
+                    attachment_id = self.env['ir.attachment'].search([('name', '=', filename)], limit=1)
+                    if attachment_id:
+                        attachment_id.update({'datas': base64.encodebytes(content)})
```
If I well understand, my ubl file is updated when sending the invoice? This is weird...

smetl commented on addons/account_invoice_import/wizard/account_invoice_send_views.xml:
```diff
@@ -0,0 +1,18 @@
+                    <div name="is_xml">
+                        <field name="is_xml" />
+                        <b><label for="is_xml"/></b>
+                    </div>
```
Is `<field name="is_xml"/>` not enough?

bfr-o replied:
No because the label is on the left of the field. The same is done for the other fields in this view

smetl commented on addons/account_invoice_import_ubl/models/account_move.py:
```diff
@@ -174,8 +179,48 @@ def _decode_ubl_2_1(self, tree):
+                    quantity=line.quantity,
+                    product=line.product_id,
+                    partner=self.partner_id)['taxes'],
+            }
```
Why not passing a list to the xml? `'invoice_line_values': [{...} for line in self.invoice_line_ids]`

smetl commented on addons/account_invoice_import_ubl/models/account_move.py:
```diff
@@ -174,8 +179,48 @@ def _decode_ubl_2_1(self, tree):
     def _get_xml_decoders(self):
-        # Override
-        ubl_decoders = [('UBL 2.1', self._detect_ubl_2_1, self._decode_ubl_2_1)]
+        # OVERRRIDE
```
Plz add morrrrrrrrrrre `R`

qdp-odoo commented on odoo/tools/pdf.py:
```diff
@@ -32,12 +44,145 @@ def rotate_pdf(pdf):
+                    'object': content_obj,
+                })
+        return files
+
```
```python
    def getAttachment(self, name):
        for i in range (0, len(self.trailer["/Root"]["/Names"]["/EmbeddedFiles"]["/Names"]), 2):
            if self.trailer["/Root"]["/Names"]["/EmbeddedFiles"]["/Names"][i+1].getObject()["/F"]==name:
                return self.trailer["/Root"]["/Names"]["/EmbeddedFiles"]["/Names"][i+1].getObject()["/EF"]["/F"].getObject().getData()

    def listAttachments(self):
        for i in range (0, len(self.trailer["/Root"]["/Names"]["/EmbeddedFiles"]["/Names"]), 2):
            yield self.trailer["/Root"]["/Names"]["/EmbeddedFiles"]["/Names"][i+1].getObject()["/F"]
```

qdp-odoo commented on odoo/tools/pdf.py:
```diff
@@ -32,12 +44,145 @@ def rotate_pdf(pdf):
+
+class OdooPdfFileWriter(PdfFileWriter):
+    # OVERRIDE of PdfFileWriter to add the management of multiple embedded files.
+
```
```python
    def addAttachment(self, fname, fdata):
        #overwrite of the AddAttachment method to allow appending attachemnts when some already exist
        if self._root_object.get('/Names') and self._root_object['/Names'].get('/EmbeddedFiles'):
            # some attachments are already there
            embeddedFilesDictionary=self._root_object["/Names"]["/EmbeddedFiles"]
            embeddedFilesNamesDictionary=embeddedFilesDictionary["/Names"]
            embeddedFilesNamesDictionary.append(self._create_attachment_object(fname, data))
        else:
            # call super
            ...
```

smetl commented on addons/account_edi/models/account_move.py:
```diff
@@ -0,0 +1,117 @@
+
+        if not self.env.context.get('no_new_invoice') and len(self) == 1 and self.state == 'draft' and (
+            self.env.context.get('default_move_type', self.move_type) in self.env['account.move'].get_invoice_types(include_receipts=True)
+            or self.env['account.journal'].browse(self.env.context.get('default_journal_id')).type in ('sale', 'purchase')
```
this condition seems really stupid.
```
if self.env.context.get('no_new_invoice'):
  return res

if len(self) != 1 or not move.is_invoice(include_receipts=True):
  return res

for attachment in self.env['ir.attachment'].browse(kwargs.get('attachment_ids', [])):
    self._create_invoice_from_attachment(attachment)
```

smetl commented on addons/account_edi/models/account_move.py:
```diff
@@ -0,0 +1,117 @@
+                move._create_invoice_from_pdf(attachment)
+        if 'xml' in attachment.mimetype:
+            for move in self:
+                move._create_invoice_from_xml_attachment(attachment)
```
shouldn't be this method `ensure_one`?

smetl commented on odoo/tools/pdf.py:
```diff
@@ -32,12 +44,78 @@ def rotate_pdf(pdf):
+    def getAttachment(self, name):
+        for i in range(0, len(self.trailer["/Root"]["/Names"]["/EmbeddedFiles"]["/Names"]), 2):
+            if self.trailer["/Root"]["/Names"]["/EmbeddedFiles"]["/Names"][i+1].getObject()["/F"]==name:
+                return self.trailer["/Root"]["/Names"]["/EmbeddedFiles"]["/Names"][i+1].getObject()["/EF"]["/F"].getObject().getData()
```
`yield`?

bfr-o replied:
We return the first attachment with the name 'name'. There shouldn't be multiple attachment with the same name.

I added documented the function to be sure.

smetl replied:
why not  `getAttachments()` that will return a dictionary <filename> -> <filecontent> instead of two methods

bfr-o replied:
More flexibility ? Since pdf.py is in tools it allows to list all files and only extract the one needed.

smetl commented on addons/account_edi/models/account_move.py:
```diff
@@ -0,0 +1,117 @@
+        if 'pdf' in attachment.mimetype:
+            self._create_invoice_from_pdf(attachment)
+        if 'xml' in attachment.mimetype:
+            self._create_invoice_from_xml_attachment(attachment)
```
this part is confusing...
`_create_invoice_from_pdf` is creating an invoice from a pdf but... what is `self` then? I don't understand...

smetl commented on addons/account_edi/views/account_journal_views.xml:
```diff
@@ -0,0 +1,15 @@
+                    </group>
+                </xpath>
+            </field>
+        </record>
```
add a `<data/>` or reduce the intendation size

smetl commented on addons/account_edi/wizard/account_invoice_send.py:
```diff
@@ -0,0 +1,36 @@
+    _description = 'Account Invoice Send'
+
+    include_xml = fields.Boolean('XML')
+    xml_ids = fields.Many2many('ir.attachment', 'account_invoice_send_ir_attachments_rel', 'wizard_id', 'attachment_id')
```
`compute='_compute_xml_ids'`

```
@api.depends('include_xml')
def _compute_xml_ids(self):
    ....
```

or remove `xml_ids` and transform `attachment_ids` as an overridable computed field

smetl commented on addons/l10n_be_edi/data/EFFF_template.xml:
```diff
@@ -0,0 +1,11 @@
+    <data>
+
+        <record id="edi_efff" model="account.edi">
+            <field name="name">EFFF</field>
```
I think it will be clearer for the end user something like `E-FFF (BE)`

smetl commented on addons/l10n_it_edi/data/invoice_it_template.xml:
```diff
@@ -1,6 +1,11 @@
+
+        <record id="edi_fatturaPA" model="account.edi">
+            <field name="name">FatturaPA</field>
+        </record>
```
Very strange to have the declaration of a new record inside a `.+template.xml` file

smetl commented on addons/l10n_it_edi/models/account_invoice.py:
```diff
@@ -254,8 +254,10 @@ def get_vat_country(vat):
         }
-        content = self.env.ref('l10n_it_edi.account_invoice_it_FatturaPA_export').render(template_values)
-        return content
+        xml_content = b"<?xml version='1.0' encoding='UTF-8'?>"
```
maybe we could add the xml header in the generic module?

bfr-o replied:
As discussed live, this would create some cluter in the code since we don't return xml_content directly but rather a dictionary with different values. It would require some additional nasty computation.

smetl commented on addons/l10n_it_edi/models/account_invoice.py:
```diff
@@ -705,8 +707,14 @@ def _compose_multi_info_message(self, tree, element_tags):
         # Override
-        ubl_decoders = [('Italy', self._detect_italy_edi, self._import_xml_invoice)]
-        return super(AccountMove, self)._get_xml_decoders() + ubl_decoders
+        if not self.journal_id.edi_ids.filtered(lambda e: e.name == 'FatturaPA'):
```
we need something more generic I think. Maybe a code on `account.edi` and retrieve the decoders based on this?
For example, `code=ubl` and make something like 

```
for edi in self.journal_id.edi_ids:
  decoders.add(getattr('_get_%s_decoder' % edi.code))
```

to discuss

bfr-o replied:
As discussed, we will use hooks in combination with `code == ubl`

smetl commented on addons/account_ubl/models/account_move.py:
```diff
@@ -1,29 +1,21 @@
 # -*- coding: utf-8 -*-
```
Maybe all the edi should be prefixed in the same way to retrieve them easily
account_edi
account_edi_facturx
account_edi_ubl?

smetl commented on addons/account_edi/models/account_edi.py:
```diff
@@ -0,0 +1,188 @@
+
+    name = fields.Char(index=True)
+    code = fields.Char()
+    embed_to_pdf = fields.Boolean(default=True)
```
Since every edi has its own implementation, maybe this field could be a method directly
```
def _embed_invoice_attachment_inside_pdf(self):
  self.ensure_one()
  return True
```

bfr-o replied:
This is not the method that check if the edi has been posted (to government for example), this is just a field stating if the file should be embed to pdf or sent has an attachment on export.

Previously, it was returned by _get_xml_decoders but since we created the model, it made more sense to put this kind of information as a field (_get_xml_decoders no longer exists).

The hook to check if an edi has been posted will be added later since it's not used in this task.

smetl replied:
something like `export_invoice = fields.Selection(selection=[('embed_to_pdf', '...'), ('separated_attachment', '...')])` to bring more flexibility maybe?

smetl commented on addons/account_edi/models/account_edi.py:
```diff
@@ -0,0 +1,188 @@
+        if 'xml' in attachment.mimetype:
+            self._create_invoice_from_xml(attachment.name, attachment.datas, True, invoice)
+        else:
+            self._create_invoice_from_file(attachment.name, attachment.datas, True, invoice)
```
In my opinion, this method should be called on records and be like this:
```
def _import_invoice_from_attachment(self, attachment):
  if not self:
    return self.env['account.move']

  if 'pdf' in attachment.mimetype:
    buffer = io.BytesIO(content)
    try:
      reader = OdooPdfFileReader(buffer)
      invoice = self._import_invoice_from_pdf_reader(attachment, reader):
      if invoice:
        return invoice
      for filename in reader.listAttachments():
        content = reader.getAttachment(filename)
        try:
            xml_tree = etree.fromstring(content)
            self._import_invoice_from_xml_tree(filename, xml_tree)
        except:
          _logger(...)
    except:
      _logger.exception("Error when reading the pdf: %s" % e)
   ...

```

smetl commented on addons/account_facturx/models/account_edi.py:
```diff
@@ -39,28 +64,39 @@ def format_monetary(number, currency):
+        :returns: the created account.move if any.
+        """
+        if code != 'facturx':
+            return super()._decode_xml(code, tree, move)
```
it should be fine to have that in two methods:

```
if code == 'facturx':
  return self._import_facturx_invoice(filename, xml_tree)
return super()....
```

bfr-o replied:
Why not keep it in one method ?

smetl replied:
It seems cleaner to me because both methods have not the same purpose and should have separated docstrings but it's just my point of view

smetl commented on addons/account_ubl/models/account_edi.py:
```diff
@@ -1,41 +1,66 @@
+
+        namespaces = _get_ubl_namespaces()
+        if not move:
+            move = self.env['account.move'].create({})
```
this should be in the generic part

bfr-o commented on addons/account_edi/models/account_edi.py:
```diff
@@ -0,0 +1,188 @@
+        return invoice
+
+    @api.model
+    def _create_invoice_from_xml(self, filename, content, is_base64=False, invoice=None):
```
Talked live

smetl commented on addons/l10n_be_edi/data/account_edi_data.xml:
```diff
@@ -0,0 +1,12 @@
+
+        <record id="edi_efff" model="account.edi">
+            <field name="name">E-FFF (BE)</field>
+            <field name="code">efff</field>
```
add version number in code

smetl commented on addons/l10n_it_edi/models/account_edi.py:
```diff
@@ -0,0 +1,427 @@
+            if multi or not invoice:
+                # make sure all the iterations create a new invoice record (except the first which could have already created one)
+                invoice = self.env['account.move']
+            multi = True
```
it seems we need to support an xml containing multiple invoices...

smetl commented on addons/l10n_it_edi/models/account_edi.py:
```diff
@@ -0,0 +1,427 @@
+        :param tree: the tree of the xml to import.
+        :param invoice: the invoice to set the decoded data.
+        :returns: the created invoice.
+        """
```
Don't need to copy/paste each time the docstring

smetl commented on addons/l10n_it_edi/models/account_edi.py:
```diff
@@ -0,0 +1,427 @@
+            # TD06 == fee
+            elements = tree.xpath('//DatiGeneraliDocumento/TipoDocumento')
+            if elements and elements[0].text and elements[0].text == 'TD01':
+                type = 'in_invoice'
```
rename to move_type

smetl commented on addons/l10n_it_edi/models/ir_mail_server.py:
```diff
@@ -140,7 +140,7 @@ def _create_invoice_from_mail(self, att_content, att_name, from_address):
             raise UserError(_('The xml file is badly formatted : {}').format(att_name))
 
-        invoice = self.env['account.move']._import_xml_invoice(tree)
+        invoice = self.env['account.edi'].search(['code', '=', 'fattura_pa'], limit=1)._import_invoice_xml_tree(tree)
```
self.env.ref('l10n_it_edi.edi_fatturaPA)

bfr-o commented on addons/account_edi/wizard/account_invoice_send.py:
```diff
@@ -0,0 +1,41 @@
+
+
+class AccountInvoiceSend(models.TransientModel):
+    _inherit = 'account.invoice.send'
```
It does. When the wizard is opened, the attachments are created, and they remain even if closing the wizard without sending. It does that for the pdf, too.

For the xmls, what we could do is generate the xmls only in the onchange, when the EDI box is ticked. The problem being that if the user untick the box, the xml are still generated (would be easy to delete, though).

It's not possible for the pdf, though.

smetl commented on addons/account/models/account_journal.py:
```diff
@@ -544,11 +544,13 @@ def create_invoice_from_attachment(self, attachment_ids=[]):
             invoice = self.env['account.move'].create({})
             attachment.write({'res_model': 'mail.compose.message'})
-            invoice.message_post(attachment_ids=[attachment.id])
+            invoice = self.env['account.edi']._create_invoice_from_attachment(attachment)
```
self.env['account.edi'].search([])?

smetl commented on addons/account_edi/models/account_edi.py:
```diff
@@ -0,0 +1,159 @@
+        :param invoice:     The invoice where to import data or None to create a new one.
+        '''
+        if invoice and len(invoice) > 1:
+            raise UserError("Invoice can only be created from attachment for one invoice")
```
I think this check is overkill

smetl commented on addons/account_edi/models/account_edi.py:
```diff
@@ -0,0 +1,159 @@
+        updated_invoice = invoice or self.env['account.move'].create({})
+
+        if 'pdf' in attachment.mimetype:
+            success = self._import_invoice_from_pdf(attachment, updated_invoice)
```
```
if invoice:
  self._update_invoice_from_pdf(attachment, invoice)
else:
  self._creates_invoice_from_pdf(attachment)
```

bfr-o commented on addons/account/models/account_journal.py:
```diff
@@ -548,7 +548,8 @@ def create_invoice_from_attachment(self, attachment_ids=[]):
             invoice = self.env['account.move'].create({})
             attachment.write({'res_model': 'mail.compose.message'})
-            invoice.message_post(attachment_ids=[attachment.id])
+            invoice = self.env['account.edi'].search([])._create_invoice_from_attachment(attachment)
```
Calling _create_invoice_from_attachment from here creates a circular dependency between account and account_edi

smetl commented on addons/account_edi/models/account_edi.py:
```diff
@@ -0,0 +1,178 @@
+        pdf_content = buffer.getvalue()
+        reader_buffer.close()
+        buffer.close()
+        return pdf_content
```
For performance reason, do this in two steps in order to create the buffer only if you have something to embed

smetl commented on addons/account_edi/models/account_edi.py:
```diff
@@ -0,0 +1,178 @@
+                attachment = self.env['ir.attachment'].create(vals)
+                attachments.append(attachment.id)
+
+        return attachments
```
do this part in batch:
```
attachment_vals_list = []
for edi in self.filtered(lambda e: e.export_invoice == 'separated_attachment'):
            vals = edi._export_invoice_to_attachment(invoice)
            if vals:
                vals['datas'] = base64.encodebytes(vals['datas'])
                attachment_vals_list.append(vals)
return self.env['ir.attachment'].create(attachment_vals_list)
```
Is there a reason to return ids instead of the records?

smetl commented on addons/account_edi/models/account_edi.py:
```diff
@@ -0,0 +1,246 @@
+        """ Create ir.attachment for the EDIs from invoice.
+
+        :param invoice: the invoice to generate the EDI from.
+        :returns: the ids of the created attachments.
```
the newly created attachments

smetl commented on addons/account_edi/models/account_edi.py:
```diff
@@ -0,0 +1,246 @@
+
+        def _decode_pdf(content):
+            try:
+                return OdooPdfFileReader(buffer)
```
buffer is undefined

bfr-o commented:
@qdp-odoo ping

qdp-odoo commented on addons/account_edi/models/account_edi.py:
```diff
@@ -0,0 +1,251 @@
+        :returns:           The invoice where to import data.
+        """
+        for file_data in self._decode_attachment(attachment):
+            for edi in self:
```
```suggestion
            for edi_format in self:
```

qdp-odoo commented on addons/account_edi/models/account_edi.py:
```diff
@@ -0,0 +1,251 @@
+                    'pdf_reader': pdf_reader,
+                })
+
+        elif 'xml' in attachment.mimetype:
```
you sure if you should do anything, but I see you don't try to unwrap a single XML attachment into several ir.attachment.

So you might want to move out the inner function `_decode_xml` to allow some XML specific format to be unwrapped as well in several files (I think the EFFF can include the PDF inside the XML, some other yet unknown format might include several invoices in a single XML doc...).

What do you think?

bfr-o replied:
I moved the _decode functions outside so we'll be able to override if necessary.

As for EFFF, the pdf is inside the XML but right it doesn't seem necessary to process it, so we're not doing it for now.

qdp-odoo commented on addons/account_edi/models/account_journal.py:
```diff
@@ -0,0 +1,18 @@
+class AccountJournal(models.Model):
+    _inherit = 'account.journal'
+
+    edi_ids = fields.Many2many(comodel_name='account.edi', string='EDI Format', help='When sending an invoice, the following XML will be attached or embedd to the PDF.')
```
```suggestion
    edi_format_ids = fields.Many2many(comodel_name='account.edi.format', string='EDI Formats', help='When sending an invoice, the following XML will be attached or embedded to the PDF.')
```

qdp-odoo commented on addons/account_edi/models/account_journal.py:
```diff
@@ -0,0 +1,18 @@
+
+    edi_ids = fields.Many2many(comodel_name='account.edi', string='EDI Format', help='When sending an invoice, the following XML will be attached or embedd to the PDF.')
+
+    def _create_invoice_from_attachment(self, attachment):
```
This method shouldn't be defined on `account.journal`, it has nothing to do with it

smetl replied:
The import from the mail alias is implemented in `account.journal`. This is not a new stuff, just an additional hook

bfr-o replied:
It's when you use the 'upload' button on a journal.

qdp-odoo commented on addons/account_edi/models/account_journal.py:
```diff
@@ -0,0 +1,18 @@
+    def _create_invoice_from_attachment(self, attachment):
+        invoice = self.env['account.edi'].search([])._create_invoice_from_attachment(attachment)
+        if invoice:
+            invoice.with_context(no_new_invoice=True).message_post(attachment_ids=[attachment.id])
```
comment context usage

qdp-odoo commented on addons/account_edi/models/account_journal.py:
```diff
@@ -0,0 +1,18 @@
+            invoice.with_context(no_new_invoice=True).message_post(attachment_ids=[attachment.id])
+            return invoice
+        else:
+            return super()._create_invoice_from_attachment(attachment)
```
``` diff
-def _create_invoice_from_attachment(self, attachment):
-        invoice = self.env['account.edi'].search([])._create_invoice_from_attachment(attachment)
-        if invoice:
-            invoice.with_context(no_new_invoice=True).message_post(attachment_ids=[attachment.id])
-            return invoice
-        else:
-            return super()._create_invoice_from_attachment(attachment)
+def _create_invoice_from_attachment(self, attachment):
+        invoice = super()._create_invoice_from_attachment(attachment)
+        self.env['account.edi.format'].search([])._import_invoice_from_attachment(attachment, invoice=invoice)
+        return invoice
```

smetl replied:
This is wrong. Some xml could contains multiple invoices. So we need to make the distinction between updating and creating new invoices. In l10n_it, for example, we don't want to update an existing invoice if the xml contains multiple invoices but we allow importing the xml directly to create multiple invoices. Then, `self.env['account.edi'].search([])._create_invoice_from_attachment(attachment)` can return multiple invoices (I agree this is not very clear in the code).

qdp-odoo replied:
forget that

qdp-odoo commented on addons/account_edi/models/account_move.py:
```diff
@@ -0,0 +1,41 @@
+
+    @api.returns('mail.message', lambda value: value.id)
+    def message_post(self, **kwargs):
+        # OVERRIDE
```
please document the reasons of the override as an introduction to the method

qdp-odoo commented on addons/account_edi/models/account_edi.py:
```diff
@@ -0,0 +1,251 @@
+                    file_data['pdf_reader'].stream.close()
+                if res:
+                    if 'extract_state' in res:
+                        res.write({'extract_state': 'done'})
```
what's that? comment.

qdp-odoo commented on addons/account_edi/models/account_move.py:
```diff
@@ -0,0 +1,41 @@
+            invoice = self.env['account.edi'].search([])._update_invoice_from_attachment(attachment, self)
+            if invoice:
+                if 'extract_state' in invoice:
+                    invoice.write({'extract_state': 'done'})
```
should be done by `_update_invoice_from_attachment` (or `_import_from_attachment`)

qdp-odoo commented on addons/account_edi/wizard/account_invoice_send_views.xml:
```diff
@@ -0,0 +1,23 @@
+                        <field name="include_edi" />
+                        <b><label for="include_edi"/></b>
+                        <div class="text-danger d-inline-block" attrs="{'invisible': ['|', ('include_edi', '=', False), ('edi_ids', '!=', [])]}">
+                            Please, add an EDI format on this invoice's journal.
```
doesn't seem right. to be tested

smetl commented on addons/account_edi/models/account_edi.py:
```diff
@@ -0,0 +1,251 @@
+
+class AccountEdi(models.Model):
+    _name = 'account.edi'
+    _description = 'EDI format'
```
I'm ok with that

smetl commented on addons/account_edi/models/account_edi.py:
```diff
@@ -0,0 +1,251 @@
+
+    name = fields.Char(index=True)
+    code = fields.Char()
+    export_invoice = fields.Selection(selection=[('embed_to_pdf', 'Embed in PDF'), ('separated_attachment', 'Attachment'), ('not_supported', 'Not Supported')], 
```
I didn't see this entry in the selection. This is the reason why you should'nt exceed 140 chars. This is more readable to me:
```
 export_invoice = fields.Selection(selection=[
  ('embed_to_pdf', 'Embed in PDF'),
  ('separated_attachment', 'Attachment'),
  ('not_supported', 'Not Supported'),
], default='embed_to_pdf')

```

bfr-o replied:
Some EDI only support import and not export.

For example, Italy, the export caused many problems so we chose not to adapt it now. Instead it is still managed internally. But we still need to adapt the import otherwise it would have broken the feature.

bfr-o replied:
I added a comment to make that  clear.

smetl commented on addons/account_edi/models/account_edi.py:
```diff
@@ -0,0 +1,251 @@
+        return False
+
+    def _update_invoice_from_xml_tree(self, filename, tree, invoice):
+        """ Import an xml into an invoice.
```
we made the distinction between `_update_blablabla` (update an existing invoice) and `_create_blablabla` (create a new invoice from scratch).

smetl commented on addons/account_edi/models/account_edi.py:
```diff
@@ -0,0 +1,251 @@
+                    return res
+        return False
+
+    def _update_invoice_from_attachment(self, attachment, invoice):
```
noooo plz, we don't want an optional parameter

smetl commented on addons/l10n_it_edi/models/account_edi.py:
```diff
@@ -0,0 +1,419 @@
+        self.ensure_one()
+        if self.code != 'fattura_pa':
+            return super()._update_invoice_from_xml_tree(filename, tree, invoice)
+        if not re.search("([A-Z]{2}[A-Za-z0-9]{2,28}_[A-Za-z0-9]{0,5}.(xml.p7m|xml))", filename):
```
Need a helper for this

smetl replied:
Also, something is missing here:
```
if multi:
  invoice.message_post("sorry but wa can't update your invoice because there is a lot of trucs in your xml")
  return False
```

smetl commented on addons/l10n_it_edi/models/account_edi.py:
```diff
@@ -0,0 +1,419 @@
+            if multi or not invoice:
+                # make sure all the iterations create a new invoice record (except the first which could have already created one)
+                invoice = self.env['account.move']
+            multi = True
```
As the multi mode is managed outside this method, the multi is simply a loot using the Form view

smetl commented on addons/account_edi/models/account_edi.py:
```diff
@@ -0,0 +1,251 @@
+            reader = OdooPdfFileReader(reader_buffer)
+            writer = OdooPdfFileWriter()
+            writer.cloneReaderDocumentRoot(reader)
+            for vals in attachements:
```
`attachments`

smetl commented on addons/account_edi_ubl/models/account_edi.py:
```diff
@@ -44,9 +47,9 @@ def _decode_ubl_2_1(self, tree):
         else:
             type = 'in_invoice'
```
rename to move_type plz

smetl commented on addons/account_edi/models/account_edi.py:
```diff
@@ -0,0 +1,251 @@
+        # TO OVERRIDE
+        self.ensure_one()
+
+        return False
```
remove the empty line plz

smetl commented on addons/account_edi/models/account_edi.py:
```diff
@@ -0,0 +1,275 @@
+        ('embed_to_pdf', 'Embed in PDF'),
+        ('separated_attachment', 'Attachment'),
+        ('not_supported', 'Not Supported')  # only import is supported
+    ], default='embed_to_pdf')
```
@qdp-odoo We could remove this field and add a method to be overridden instead. What do you think? It could add more flexibility

smetl commented on addons/account_edi/models/account_edi.py:
```diff
@@ -0,0 +1,275 @@
+            buffer.close()
+        return pdf_content
+
+    def _create_ir_attachments(self, invoice):
```
We need a way to decide when the xml will be generated. This method is currently called when opening the "send by mail" wizard but some EDI will need to generate the xml when posting the invoice. What about having a method instead the field on account.edi.format like:
```
def _export_invoice_options(self):
  return {
    'export_type': 'embed_to_pdf' / 'separated_file',
    'export_generation': 'at_post' / 'at_send',
  }
```

smetl commented on addons/account_edi/models/account_edi.py:
```diff
@@ -0,0 +1,275 @@
+        """ Create the file content representing the invoice.
+
+        :param invoice: the invoice to encode.
+        :returns: a dictionnary (values are compatible to create an ir.attachment)
```
typo: `dictionnary` must be `dictionary`

smetl commented on addons/account_edi/models/account_edi.py:
```diff
@@ -0,0 +1,275 @@
+
+
+class AccountEdiFormat(models.Model):
+    _name = 'account.edi.format'
```
rename the file into account_edi_format.py plz

smetl commented on addons/account_edi/models/account_journal.py:
```diff
@@ -0,0 +1,21 @@
+    _inherit = 'account.journal'
+
+    edi_format_ids = fields.Many2many(comodel_name='account.edi.format',
+                                      string='Electronic invoicing',
```
`domain="[('can_export', '=', True)]"` ?

smetl commented on addons/account_edi/models/account_edi_format.py:
```diff
@@ -0,0 +1,307 @@
+        :returns: True if the invoice was successfully.
+        """
+        # TO OVERRIDE
+        return True  # move without edi are posted automatically
```
This method should be added with the intermediate states

qdp-odoo replied:
this method is still to be removed from this PR

qdp-odoo commented on addons/account_edi_facturx/models/account_edi_format.py:
```diff
@@ -39,35 +38,59 @@ def format_monetary(number, currency):
+    def _create_invoice_from_xml_tree(self, filename, tree):
+        self.ensure_one()
+        if self.code == 'facturx_1_0_05':
+            if tree.tag == '{urn:un:unece:uncefact:data:standard:CrossIndustryInvoice:100}CrossIndustryInvoice':
```
factorize those 2 lines in a function called here and in next function

bfr-o commented on addons/account_edi/models/account_edi_format.py:
```diff
@@ -0,0 +1,307 @@
+    _name = 'account.edi.format'
+    _description = 'EDI format'
+
+    name = fields.Char(index=True)
```
Yes, it's not necessary. I removed it

bfr-o commented on addons/account_edi/models/account_move.py:
```diff
@@ -0,0 +1,50 @@
+                              author_id=odoobot.id)
+            return res
+        if attachments and self.invoice_line_ids:
+            self.message_post(body='The invoice already contains lines, it was not be updated from the attachment.',
```
Yes, sorry :)

bfr-o commented on addons/account_edi/wizard/account_invoice_send.py:
```diff
@@ -0,0 +1,32 @@
+            requested_formats = r.edi_format_ids
+            for invoice in r.invoice_ids:
+                available_formats = existing_attachments.filtered(lambda a: a.res_id == invoice._origin.id).edi_format_id
+                missing_formats = requested_formats.filtered(lambda f: f._origin.id not in available_formats.ids)
```
I'm not sure, strictly speaking, but if I don't do that, it doesn't work.

- For available_formats : r.invoice_ids.ids is a list of <NewId> so the comparison with res_id always fails.
- For missing_formats : requested formats is also a list of <NewId> whereas available_formats is ids (int). Probably because r.edi_format_ids is a many2many in a TransientModel so I'm guessing the m2m is only in cache.

bfr-o commented on addons/l10n_be_edi/models/account_move.py:
```diff
@@ -0,0 +1,15 @@
+    def _get_ubl_values(self):
+        values = super(AccountMove, self)._get_ubl_values()
+
+        # E-fff uses ubl_version 2.0, the UBL we generate doesn't use the feature of 2.1, so we can use the 2.1 generator
```
I rephrased

bfr-o commented on addons/account_edi/models/account_edi_format.py:
```diff
@@ -0,0 +1,307 @@
+            This hook allows to customize the EDI format generation based on the pdf.
+            For example : to include the pdf inside the EDI format.
+            The default behavior is to call _export_invoice_to_attachment.
+            Should return False for EDI format that must not be embedded to pdf. 
```
Tried to make it better

qdp-odoo commented on addons/account_edi/models/account_edi_format.py:
```diff
@@ -0,0 +1,307 @@
+        """
+        # TO OVERRIDE
+        self.ensure_one()
+        return self._export_invoice_to_attachment(invoice)
```
ho I get i! I was suggesting to return False, by default.
 
But no, what we want is
* the default is to add the EDI document inside the PDF, because we find it convenient 
* but one may want not to do it (and return False instead)
* but one may need to do some special treatment before and overwrite the behavior

Was it clear enough from the docstring?

qdp-odoo commented on addons/account_edi/wizard/account_invoice_send.py:
```diff
@@ -0,0 +1,32 @@
+            r.edi_attachment_ids = existing_attachments
+
+            requested_formats = r.edi_format_ids
+            for invoice in r.invoice_ids:
```
I mean: I'm not sure what this loop does, but I'm sure of what it shouldn't do!

It should not attach the EDI documents of all selected invoices and send them to the different customers (yeah, different invoices mean potentialy different customers)

bfr-o commented:
@qdp-odoo I think I addressed all your comments.

Concerning l10n_it_einvoice_name, l10n_it_einvoice_id. These are the attachment, and are generated in post. I think it should stay that way because a compute would be complicated (lots of dependencies) and inefficient (regenerating the xml each time something is changed in the invoice).

qdp-odoo commented on addons/account_edi/models/mail_template.py:
```diff
@@ -0,0 +1,33 @@
+    _inherit = "mail.template"
+
+    def generate_email(self, res_ids, fields):
+        print('generate_email ici')
```
oops

qdp-odoo commented on addons/account_edi_facturx/models/account_edi_format.py:
```diff
@@ -39,35 +38,59 @@ def format_monetary(number, currency):
+                }
+
+    def _should_import(self, filename, tree):
+        return self.code == 'facturx_1_0_05' and tree.tag == '{urn:un:unece:uncefact:data:standard:CrossIndustryInvoice:100}CrossIndustryInvoice'
```
bad inheritance: _should_import method defined twice without call to super()

qdp-odoo commented:
> @qdp-odoo I think I addressed all your comments.
> 
> Concerning l10n_it_einvoice_name, l10n_it_einvoice_id. These are the attachment, and are generated in post. I think it should stay that way because a compute would be complicated (lots of dependencies) and inefficient (regenerating the xml each time something is changed in the invoice).

I think the `account_edi` module should add a field `edi_document_ids` on `account.move` (One2Many ir.attachment). That one doesn't need to be computed.

Then you could have a simple computed field `l10n_it_einvoice_id` on the first `ir.attachment `having edit_format = it_edi.

qdp-odoo commented on addons/account_edi/models/account_move.py:
```diff
@@ -0,0 +1,53 @@
+class AccountMove(models.Model):
+    _inherit = 'account.move'
+
+    edi_document_ids = fields.One2many('ir.attachment', 'res_id', domain=[('res_model', '=', 'account.move')])
```
```suggestion
    edi_document_ids = fields.One2many('ir.attachment', 'res_id', domain=[('res_model', '=', 'account.move'), ('edi_format_id', '!=', False)])
```

smetl commented on addons/account_edi/models/account_edi_format.py:
```diff
@@ -0,0 +1,303 @@
+        for edi_format in self:
+            try:
+                vals = edi_format._export_invoice_to_attachment(invoice)
+            except:
```
Why do you need this `try: except:` ?

smetl commented on addons/account_edi/models/account_move.py:
```diff
@@ -0,0 +1,53 @@
+                              subtype_xmlid='mail.mt_note',
+                              author_id=odoobot.id)
+            return res
+        if attachments and self.invoice_line_ids:
```
`self.line_ids` to be sure

qdp-odoo commented:
@robodoo rebase-merge r+

robodoo commented:
Linked pull request(s) odoo/upgrade#1108 not ready. Linked PRs are not staged until all of them are ready.

qdp-odoo commented:
@robodoo r+ rebase-merge

robodoo commented:
Merge method set to rebase and merge, using the PR as merge commit message

robodoo commented:
Staging failed: ci/runbot on f7dc4f6b4eed489fc1c3107e0f310d22ba384d7b (view more at http://runbot.odoo.com/runbot/build/2867496)

d-fence commented:
robodoo retry
